File: firmware/src/particles.cpp
#include "particles.h"
#include "gauge.h"

void ParticleSystem::init(const ParticleConfig& cfg) {
    config = cfg;
    for (auto& p : particles) p.alive = false;
    for (auto& em : config.emitters) em.accumulator = 0;

    memset(maskGrid, 0, sizeof(maskGrid));
    if (config.mask.length() >= MATRIX_WIDTH * MATRIX_HEIGHT) {
        for (uint8_t y = 0; y < MATRIX_HEIGHT; y++) {
            for (uint8_t x = 0; x < MATRIX_WIDTH; x++) {
                maskGrid[x][y] = config.mask[y * MATRIX_WIDTH + x] == '#';
            }
        }
    }
}

bool ParticleSystem::isMaskSolid(int16_t x, int16_t y) {
    if (x < 0 || x >= MATRIX_WIDTH || y < 0 || y >= MATRIX_HEIGHT) return false;
    return maskGrid[x][y];
}

uint32_t ParticleSystem::randomColor() {
    if (config.colors.stops.empty()) return 0xFFFFFF;
    float pos = random(0, 1000) / 1000.0f;
    float val = config.colors.min_val + pos * (config.colors.max_val - config.colors.min_val);
    return colorFromRange(config.colors, val);
}

void ParticleSystem::spawn(const ParticleEmitter& em) {
    for (auto& p : particles) {
        if (p.alive) continue;
        p.alive = true;
        p.x = em.x < 0 ? random(0, MATRIX_WIDTH * 100) / 100.0f : em.x;
        p.y = em.y < 0 ? random(0, MATRIX_HEIGHT * 100) / 100.0f : em.y;
        p.vx = em.vx_min + (random(0, 1000) / 1000.0f) * (em.vx_max - em.vx_min);
        p.vy = em.vy_min + (random(0, 1000) / 1000.0f) * (em.vy_max - em.vy_min);
        p.lifetime = em.lifetime_min + random(0, em.lifetime_max - em.lifetime_min + 1);
        p.age = 0;
        p.size = em.size;
        p.color = em.is_rocket ? 0xFFFFFF : randomColor();
        p.isRocket = em.is_rocket;
        return;
    }
}

void ParticleSystem::burst(float x, float y) {
    int count = 8 + random(0, 6);
    uint32_t burstColor = randomColor();
    for (int i = 0; i < count; i++) {
        for (auto& p : particles) {
            if (p.alive) continue;
            p.alive = true;
            p.x = x;
            p.y = y;
            float angle = (float)i / count * 6.2832f + (random(0, 100) / 100.0f * 0.3f);
            float speed = 4.0f + random(0, 300) / 100.0f;
            p.vx = cos(angle) * speed * 2.0f;
            // Boost upward particles to counteract gravity, making a rounder burst
            float vy = sin(angle) * speed;
            if (vy < 0) vy *= 1.8f;  // upward gets extra kick
            p.vy = vy;
            p.lifetime = 500 + random(0, 400);
            p.age = 0;
            p.size = 1;
            p.color = burstColor;
            p.isRocket = false;
            break;
        }
    }
}

void ParticleSystem::tick(uint32_t dt_ms) {
    if (dt_ms == 0 || dt_ms > 200) dt_ms = 20;
    float dt = dt_ms / 1000.0f;

    // Spawn from emitters
    for (auto& em : config.emitters) {
        em.accumulator += em.rate * dt;
        while (em.accumulator >= 1.0f) {
            spawn(em);
            em.accumulator -= 1.0f;
        }
    }

    // Particle-particle collision (for bouncing mode with size >= 2)
    if (config.edge == PB_BOUNCE) {
        for (int i = 0; i < MAX_PARTICLES; i++) {
            if (!particles[i].alive || particles[i].size < 2) continue;
            for (int j = i + 1; j < MAX_PARTICLES; j++) {
                if (!particles[j].alive || particles[j].size < 2) continue;
                float dx = particles[j].x - particles[i].x;
                float dy = particles[j].y - particles[i].y;
                float dist = dx * dx + dy * dy;
                // Collision radius ~2px
                if (dist < 4.0f && dist > 0.01f) {
                    // Swap velocities (elastic)
                    float tvx = particles[i].vx;
                    float tvy = particles[i].vy;
                    particles[i].vx = particles[j].vx;
                    particles[i].vy = particles[j].vy;
                    particles[j].vx = tvx;
                    particles[j].vy = tvy;
                    // Push apart
                    float d = sqrtf(dist);
                    float overlap = (2.0f - d) * 0.5f;
                    float nx = dx / d;
                    float ny = dy / d;
                    particles[i].x -= nx * overlap;
                    particles[i].y -= ny * overlap;
                    particles[j].x += nx * overlap;
                    particles[j].y += ny * overlap;
                }
            }
        }
    }

    // Update particles
    for (auto& p : particles) {
        if (!p.alive) continue;

        p.age += dt_ms;
        if (p.age >= p.lifetime) {
            p.alive = false;
            continue;
        }

        float oldVy = p.vy;

        // Apply gravity
        p.vy += config.gravity * dt;

        // Rocket: burst at apex (vy crosses from negative to positive)
        if (p.isRocket && oldVy < 0 && p.vy >= 0) {
            burst(p.x, p.y);
            p.alive = false;
            continue;
        }

        // Move
        float nx = p.x + p.vx * dt;
        float ny = p.y + p.vy * dt;

        // Mask collision
        int16_t ix = (int16_t)nx;
        int16_t iy = (int16_t)ny;
        if (isMaskSolid(ix, iy)) {
            if (config.edge == PB_DIE) { p.alive = false; continue; }
            if (isMaskSolid((int16_t)nx, (int16_t)p.y)) { p.vx = -p.vx * 0.7f; nx = p.x; }
            if (isMaskSolid((int16_t)p.x, (int16_t)ny)) { p.vy = -p.vy * 0.7f; ny = p.y; }
        }

        // Edge handling
        if (config.edge == PB_BOUNCE) {
            if (nx < 0) { nx = 0; p.vx = -p.vx * 0.8f; }
            if (nx >= MATRIX_WIDTH - 1) { nx = MATRIX_WIDTH - 1.01f; p.vx = -p.vx * 0.8f; }
            if (ny < 0) { ny = 0; p.vy = -p.vy * 0.8f; }
            if (ny >= MATRIX_HEIGHT - 1) { ny = MATRIX_HEIGHT - 1.01f; p.vy = -p.vy * 0.8f; }
        } else if (config.edge == PB_WRAP) {
            if (nx < 0) nx += MATRIX_WIDTH;
            if (nx >= MATRIX_WIDTH) nx -= MATRIX_WIDTH;
            if (ny < 0) ny += MATRIX_HEIGHT;
            if (ny >= MATRIX_HEIGHT) ny -= MATRIX_HEIGHT;
        } else { // PB_DIE
            if (nx < -2 || nx >= MATRIX_WIDTH + 2 || ny < -2 || ny >= MATRIX_HEIGHT + 2) {
                p.alive = false;
                continue;
            }
        }

        p.x = nx;
        p.y = ny;
    }
}

void ParticleSystem::render(Display& display) {
    // Draw mask
    for (uint8_t y = 0; y < MATRIX_HEIGHT; y++) {
        for (uint8_t x = 0; x < MATRIX_WIDTH; x++) {
            if (maskGrid[x][y]) display.drawPixel(x, y, 0x222222);
        }
    }

    // Draw particles
    for (auto& p : particles) {
        if (!p.alive) continue;

        uint32_t color = p.color;

        // Fade out in last 25% of life
        float progress = (float)p.age / p.lifetime;
        if (progress > 0.75f) {
            float fade = 1.0f - (progress - 0.75f) * 4.0f;
            uint8_t r = ((color >> 16) & 0xFF) * fade;
            uint8_t g = ((color >> 8) & 0xFF) * fade;
            uint8_t b = (color & 0xFF) * fade;
            color = ((uint32_t)r << 16) | ((uint32_t)g << 8) | b;
        }

        int16_t px = (int16_t)p.x;
        int16_t py = (int16_t)p.y;

        if (px >= 0 && px < MATRIX_WIDTH && py >= 0 && py < MATRIX_HEIGHT) {
            display.drawPixel(px, py, color);
        }
        if (p.size >= 2) {
            // 2x2 block
            if (px + 1 < MATRIX_WIDTH && py >= 0 && py < MATRIX_HEIGHT)
                display.drawPixel(px + 1, py, color);
            if (py + 1 < MATRIX_HEIGHT && px >= 0 && px < MATRIX_WIDTH)
                display.drawPixel(px, py + 1, color);
            if (px + 1 < MATRIX_WIDTH && py + 1 < MATRIX_HEIGHT)
                display.drawPixel(px + 1, py + 1, color);
        }
    }
}

File: src/utils/renderer/particles.js
/**
 * Particle system — port of firmware particle physics.
 * @module utils/renderer/particles
 */

import { setPixel, addPixel, colorFromStops, WIDTH, HEIGHT } from './core.js';

const MAX_PARTICLES = 48;

/** Initialize particle state for a layer */
export function initParticles(layer) {
  const particles = [];
  for (let i = 0; i < MAX_PARTICLES; i++) {
    particles.push({ alive: false, x: 0, y: 0, vx: 0, vy: 0, age: 0, lifetime: 0, size: 1, color: [255,255,255] });
  }
  const state = { particles, emitters: (layer.emitters || []).map(e => ({ ...e, acc: 0 })) };
  // Warmup
  const warmup = layer.warmup || 0;
  if (warmup > 0) {
    const steps = Math.floor(warmup / 20);
    for (let i = 0; i < steps; i++) tickParticles(state, layer, 20);
  }
  return state;
}

function randomColor(colors) {
  if (!colors?.stops?.length) return [255, 255, 255];
  const pos = Math.random();
  const val = (colors.min || 0) + pos * ((colors.max || 1) - (colors.min || 0));
  return colorFromStops(colors.stops, val);
}

function spawn(state, em, colors) {
  for (const p of state.particles) {
    if (p.alive) continue;
    p.alive = true;
    p.x = em.x < 0 ? Math.random() * WIDTH : em.x;
    p.y = em.y < 0 ? Math.random() * HEIGHT : em.y;
    p.vx = em.vx_min + Math.random() * (em.vx_max - em.vx_min);
    p.vy = em.vy_min + Math.random() * (em.vy_max - em.vy_min);
    p.lifetime = em.life_min + Math.random() * (em.life_max - em.life_min);
    p.age = 0;
    p.size = em.size || 1;
    p.color = randomColor(colors);
    return;
  }
}

/** Advance particle simulation by dt_ms */
export function tickParticles(state, layer, dt_ms = 16) {
  const dt = dt_ms / 1000;
  const gravity = layer.gravity || 0;
  const edge = layer.edge || 'die';

  for (const em of state.emitters) {
    em.acc += (em.rate || 1) * dt;
    while (em.acc >= 1) {
      spawn(state, em, layer.colors);
      em.acc -= 1;
    }
  }

  for (const p of state.particles) {
    if (!p.alive) continue;
    p.age += dt_ms;
    if (p.age >= p.lifetime) { p.alive = false; continue; }
    p.vy += gravity * dt;
    p.x += p.vx * dt;
    p.y += p.vy * dt;

    if (edge === 'bounce') {
      if (p.x < 0) { p.x = 0; p.vx *= -0.8; }
      if (p.x >= WIDTH - 1) { p.x = WIDTH - 1.01; p.vx *= -0.8; }
      if (p.y < 0) { p.y = 0; p.vy *= -0.8; }
      if (p.y >= HEIGHT - 1) { p.y = HEIGHT - 1.01; p.vy *= -0.8; }
    } else if (edge === 'wrap') {
      if (p.x < 0) p.x += WIDTH;
      if (p.x >= WIDTH) p.x -= WIDTH;
      if (p.y < 0) p.y += HEIGHT;
      if (p.y >= HEIGHT) p.y -= HEIGHT;
    } else {
      if (p.x < -2 || p.x >= WIDTH + 2 || p.y < -2 || p.y >= HEIGHT + 2) {
        p.alive = false;
      }
    }
  }
}

/** Render particles to buffer */
export function renderParticles(buf, layer, _icons, state) {
  if (!state._particles) state._particles = initParticles(layer);
  tickParticles(state._particles, layer);

  const blend = layer.blend === 'add' ? addPixel : setPixel;
  for (const p of state._particles.particles) {
    if (!p.alive) continue;
    let [r, g, b] = p.color;
    const progress = p.age / p.lifetime;
    if (progress > 0.75) {
      const fade = 1 - (progress - 0.75) * 4;
      r = Math.round(r * fade);
      g = Math.round(g * fade);
      b = Math.round(b * fade);
    }
    const px = Math.floor(p.x), py = Math.floor(p.y);
    blend(buf, px, py, r, g, b);
    if (p.size >= 2) {
      blend(buf, px + 1, py, r, g, b);
      blend(buf, px, py + 1, r, g, b);
      blend(buf, px + 1, py + 1, r, g, b);
    }
  }
}
